### PR TITLE
➕ transpile-web: font support

### PR DESCRIPTION
**Pull request by xaviervia**
Only missing part is `font.shadow`, which will come from https://github.com/batata-frita/lada/pull/7

Also: made some changes to the API to simplify. Will annotate those as comments.

### CODE REVIEW

File: packages/transpile-web/src/index.js [hunk 1]
```diff
@@ -74,6 +74,29 @@ const withContent = (
 
 const withCursor = (cursor) => style('cursor', isDef(cursor), () => cursor)
 
+const withFont = (font) => compose(
+  style('whiteSpace', isDef(font) && font.shouldPreserveAllWhitespace, () => 'pre'),
+  style('whiteSpace', isDef(font), () => 'pre-line'),
```
Comment (xaviervia) [inline packages/transpile-web/src/index.js]: For any Text components, the default behavior of text is to collapse all whitespace _except_ line breaks. I find this significantly more natural than the default CSS behavior.

File: packages/transpile-web/src/index.js [hunk 2]
```diff
@@ -74,6 +74,29 @@ const withContent = (
...
+  style('color', isDef(font) && isDef(font.color), () => font.color),
+  style('display', isDef(font), () => 'inline-block'),
```
Comment (xaviervia) [inline packages/transpile-web/src/index.js]: Any Text component needs to be set to `display: inline-block` to work properly.

File: packages/transpile-web/demo/index.jsx [hunk 3]
```diff
@@ -164,6 +189,121 @@ const Demo = () => (
...
+                }
+              })}>A rather</span><br />
+            <span style={lada({
+                font: {
+                  overline: true
+                }
+              })}>clear</span>{' '}
+            <span style={lada({
+                font: {
+                  strikethrough: true
```
Comment (xaviervia) [inline packages/transpile-web/demo/index.jsx]: I changed from `decoration: underline` `decoration: overline` for just 
`underline: true` and so on. The `decoration` thing is pointless.

File: packages/transpile-web/demo/index.jsx [hunk 4]
```diff
@@ -164,6 +189,121 @@ const Demo = () => (
...
+              left: 20,
+              right: 20,
+              bottom: 30
+            }
+          })}>
+          <span style={lada({
+              font: {
+                color: 'lightgreen',
+                family: ['Menlo', 'Consolas'],
+                shouldPreserveAllWhitespace: true
```
Comment (xaviervia) [inline packages/transpile-web/demo/index.jsx]: This sets the text to `white-space: pre`. It is the only other setting that I think it's necessary to support.

**Comment by deepsweet** [comment]
Btw haven't we decided to split "layout" and "text" styles entirely? like `createLayoutStyles()` and `createTextStyles()` methods to force you to think about it more :)

File: packages/transpile-web/demo/index.jsx [hunk 5]
```diff
@@ -164,6 +189,121 @@ const Demo = () => (
...
+                lineHeight: 40,
+                weight: 300,
+                letterSpacing: 3,
+                case: 'upper',
+                align: 'center'
+              }
+            })}>
+            <span style={lada({
+                font: {
+                  underline: true
```
[the remark this answers is not in the document]
Comment (xaviervia) [reply packages/transpile-web/demo/index.jsx]: Sounds weird. What about `hasUnderline`?

[the remark this answers is not in the document]
Comment (deepsweet) [reply packages/transpile-web/demo/index.jsx]: :+1:

File: packages/transpile-web/demo/index.jsx (hunk 3 above)
Comment (deepsweet) [reply packages/transpile-web/demo/index.jsx]: same here for some boolean name.

File: packages/transpile-web/src/index.js [hunk 6]
```diff
@@ -74,6 +74,29 @@ const withContent = (
 
 const withCursor = (cursor) => style('cursor', isDef(cursor), () => cursor)
 
+const withFont = (font) => compose(
```
[the remark this answers is not in the document]
Comment (xaviervia) [reply packages/transpile-web/src/index.js]: Yes... the problem I am trying to address here is setting the defaults for any `font` tag: namely, `display: inline-block` and `white-space: pre-line`.

On a second thought, the real problem is that I'm still using the same function for layout and font. I think I should refactor that first.

**Comment by xaviervia** [comment]
As FYI, I’m working on splitting them. It makes the whole approach for font a lot easier.

**Comment by deepsweet** [comment]
I was thinking about something like:

**Comment by deepsweet** [comment]
```transforms/

**Comment by deepsweet** [comment]
omg, I can't even use my mobile github app properly for comments :)

```
transforms/
  layout.js
  text.js
index.js
utils.js
```

**Comment by xaviervia** [comment]
Just pushed some WIP so that the direction in which I’m going is clear.

**Comment by xaviervia** [comment]
OK, the refactor is complete. I did two renames:

1. `case: 'upper'` is now `isUppercase`, same for `isLowercase` and `isCapitalized`
2. `underline` is `hasUnderline`. Do you prefer `isUnderlined`?

File: packages/transpile-web/demo/index.jsx [hunk 7]
```diff
@@ -164,6 +197,120 @@ const Demo = () => (
...
+        <Paragraph>
+          <span
+            style={{
+              ...layout({
+                size: {
+                  width: 255,
+                  height: 20
+                }
+              }),
+              ...text({
```
Comment (xaviervia) [inline packages/transpile-web/demo/index.jsx]: This part I’m not sure about. Maybe `text` needs to support sizing properties. To demonstrate ellipsis and clip I needed to add fixed sizes. Wdyt, should we add support for `size` in text?

Comment (deepsweet) [reply packages/transpile-web/demo/index.jsx]: I'm all hands up for a strong separation of layout/text concepts so they should be really independent. Let's copypaste `size` there.

**Comment by xaviervia** [comment]
I’m doing the change of adding `size`, but I'm renaming it to `boundaries`.

There are two reasons:

1. We already have a `size` key for text. It’s for the `font-size``
2. This is not the size of the text, rather the boundaries of the text "box".

Makes sense?

**Comment by deepsweet** [comment]
`dimensions`?

File: packages/transpile-web/src/transforms/text.js [hunk 8]
```diff
@@ -13,6 +13,13 @@ const defaults = compose(
 
 const withAlign = (align) => style('textAlign', isDef(align), () => align)
 
+const withDimensions = ({ height, width } = {}) => compose(
+  style('overflow', isDef(height), () => 'hidden'),
+  style('overflow', isDef(width), () => 'hidden'),
```
Comment (deepsweet) [inline packages/transpile-web/src/transforms/text.js]: :heart:

**Pull request merged**